### PR TITLE
Add informative text about raising an exception…

### LINKED ISSUE #186: Algorithm interface vs DoS defense
https://www.w3.org/TR/rdf-canon/#canon-algorithm says

> The canonicalization algorithm converts an [input dataset](https://www.w3.org/TR/rdf-canon/#dfn-input-dataset) into a [canonicalized dataset](https://www.w3.org/TR/rdf-canon/#dfn-canonicalized-dataset).

However, https://www.w3.org/TR/rdf-canon/#canon-algo-algo says

> Implementations MUST prevent against potential denial-of-service attacks.

The interface claims that they're still going to produce a canonicalized dataset after detecting a denial-of-service attack, but the algorithm doesn't say anything about how they're supposed to do that. I suspect the interface should change to say that the algorithm returns either a canonicalized dataset or a failure.

### PR DESCRIPTION
**Pull request by gkellogg**
… and terminating early if a poison dataset is detected.

Fixes #186.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-canon/pull/189.html" title="Last updated on Nov 28, 2023, 10:37 PM UTC (d38d31a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-canon/189/be6773f...d38d31a.html" title="Last updated on Nov 28, 2023, 10:37 PM UTC (d38d31a)">Diff</a>

### CODE REVIEW

File: spec/index.html [hunk 1]
```diff
@@ -658,7 +658,9 @@ <h2>Blank Node Identifier Issuer State</h2>
     <h2>Canonicalization Algorithm</h2>
 
     <p>The canonicalization algorithm converts an <a>input dataset</a>
-      into a <a>canonicalized dataset</a>. This algorithm will assign
+      into a <a>canonicalized dataset</a> or raise an error if
```
Comment (TallTed) [inline spec/index.html:661]:
```suggestion
      into a <a>canonicalized dataset</a> or raises an error if
```

File: spec/index.html [hunk 2]
```diff
@@ -1055,7 +1057,8 @@ <h3>Algorithm</h3>
       <p>The following algorithm will run with a minimal number of iterations in each step
         for typical <a>input datasets</a>.
         In some extreme cases, the algorithm can behave poorly, particularly in <a href="#ca.5">Step 5</a>.
-        Implementations MUST prevent against potential denial-of-service attacks.
+        Implementations MUST prevent against potential denial-of-service attacks
+        by raising a suitable exception and terminating early.
```
Comment (TallTed) [inline spec/index.html:1061]:
```suggestion
        by raising suitable exceptions and terminating early.
```

**Review by TallTed — CHANGES_REQUESTED** [review]
matching plural/singular things

File: spec/index.html [hunk 3]
```diff
@@ -1055,7 +1057,8 @@ <h3>Algorithm</h3>
       <p>The following algorithm will run with a minimal number of iterations in each step
         for typical <a>input datasets</a>.
         In some extreme cases, the algorithm can behave poorly, particularly in <a href="#ca.5">Step 5</a>.
-        Implementations MUST prevent against potential denial-of-service attacks.
+        Implementations MUST prevent against potential denial-of-service attacks
```
Comment (jyasskin) [inline spec/index.html:1060]: This predates this PR, but something's wrong with "prevent against".
```suggestion
        Implementations MUST defend against potential denial-of-service attacks
```

**Review by jyasskin — APPROVED** [review]
Thanks!

**Review by dlongley — APPROVED** [review]

**Review by yamdan — APPROVED** [review]
Thanks!

**Review by iherman — APPROVED** [review]

**Comment by iherman** [comment]
The echidna action file must be updated before merge: "WD" -> "CRD"

**Comment by TallTed** [comment]
>> [@iherman] The echidna action file must be updated before merge: "WD" -> "CRD"

> Echidna Auto-publish WD

Seems to say the echidna action file has not been updated?

**Comment by gkellogg** [comment]
That was something @pchampin was going to look into.

**Comment by deniak** [comment]
That said, trying to publish a WD from a CRS will be rejected by echidna.

**Comment by pchampin** [comment]
@ivan @gkellogg 
As pointed out by @deniak, Echidna is currently disabled (by commenting out the token).
The "page build and deployment" action still runs he checks, and skips the publishing on TR.
We can still publish new Editors' Drafts, but they are now out of sync with TR.

I hope the action will kick-in again when I update the Echidna config, otherwise, I believe I can re-run it manually.

**Comment by iherman** [comment]
Ah! I missed the commenting trick. I usually rename the yml file in similar circumstances...

**Comment by pchampin** [comment]
I have updated the Echidna config, and I had to make a dummy commit to spec/index.html to get the action to run, but we now have a new CR draft : https://www.w3.org/TR/2023/CRD-rdf-canon-20231130/

**Pull request merged**